eslint/babel-eslint-parser/src/client.cjs

```diff
@@ -1,67 +1,105 @@
 const path = require("path");
 
-let send;
-
-exports.getVersion = sendCached("GET_VERSION");
-
-exports.getTypesInfo = sendCached("GET_TYPES_INFO");
-
-exports.getVisitorKeys = sendCached("GET_VISITOR_KEYS");
-
-exports.getTokLabels = sendCached("GET_TOKEN_LABELS");
-
-exports.maybeParse = (code, options) => send("MAYBE_PARSE", { code, options });
-
-function sendCached(action) {
-  let cache = null;
-
-  return () => {
-    if (!cache) cache = send(action, undefined);
-    return cache;
-  };
+const ACTIONS = {
+  GET_VERSION: "GET_VERSION",
+  GET_TYPES_INFO: "GET_TYPES_INFO",
+  GET_VISITOR_KEYS: "GET_VISITOR_KEYS",
+  GET_TOKEN_LABELS: "GET_TOKEN_LABELS",
+  MAYBE_PARSE: "MAYBE_PARSE",
+  MAYBE_PARSE_SYNC: "MAYBE_PARSE_SYNC",
+};
+
+class Client {
+  #send;
+
+  constructor(send) {
+    this.#send = send;
+  }
+
+  #vCache;
+  getVersion() {
+    return (this.#vCache ??= this.#send(ACTIONS.GET_VERSION, undefined));
+  }
+
+  #tiCache;
+  getTypesInfo() {
+    return (this.#tiCache ??= this.#send(ACTIONS.GET_TYPES_INFO, undefined));
+  }
+
+  #vkCache;
+  getVisitorKeys() {
+    return (this.#vkCache ??= this.#send(ACTIONS.GET_VISITOR_KEYS, undefined));
+  }
+
+  #tlCache;
+  getTokLabels() {
+    return (this.#tlCache ??= this.#send(ACTIONS.GET_TOKEN_LABELS, undefined));
+  }
+
+  maybeParse(code, options) {
+    return this.#send(ACTIONS.MAYBE_PARSE, { code, options });
+  }
 }
 
-if (process.env.BABEL_8_BREAKING) {
-  const {
-    Worker,
-    receiveMessageOnPort,
-    MessageChannel,
-    SHARE_ENV,
-  } = require("worker_threads");
-
-  // We need to run Babel in a worker for two reasons:
-  // 1. ESLint workers must be CJS files, and this is a problem
-  //    since Babel 8+ uses native ESM
-  // 2. ESLint parsers must run synchronously, but many steps
-  //    of Babel's config loading (which is done for each file)
-  //    can be asynchronous
-  // If ESLint starts supporting async parsers, we can move
-  // everything back to the main thread.
-  const worker = new Worker(
+// We need to run Babel in a worker for two reasons:
+// 1. ESLint workers must be CJS files, and this is a problem
+//    since Babel 8+ uses native ESM
+// 2. ESLint parsers must run synchronously, but many steps
+//    of Babel's config loading (which is done for each file)
+//    can be asynchronous
+// If ESLint starts supporting async parsers, we can move
+// everything back to the main thread.
+exports.WorkerClient = class WorkerClient extends Client {
+  static #worker_threads_cache;
+  static get #worker_threads() {
+    return (WorkerClient.#worker_threads_cache ??= require("worker_threads"));
+  }
```

Comment:
Q: What's the intention of private static getter? Doesn't `static #worker_threads = require("worker_threads")` work here?

Reply:
This file is always executed, but `require("worker_threads")` is only necessary when actually using the experimental implementation.